sleep.1
.\" $MirOS: src/bin/sleep/sleep.1,v 1.3 2021/01/23 03:36:59 tg Exp $
.\"-
.\" Copyright © 2021
.\"      mirabilos <m@mirbsd.org>
.\"
.\" Provided that these terms and disclaimer and all copyright notices
.\" are retained or reproduced in an accompanying document, permission
.\" is granted to deal in this work without restriction, including un‐
.\" limited rights to use, publicly perform, distribute, sell, modify,
.\" merge, give away, or sublicence.
.\"
.\" This work is provided “AS IS” and WITHOUT WARRANTY of any kind, to
.\" the utmost extent permitted by applicable law, neither express nor
.\" implied; without malicious intent or gross negligence. In no event
.\" may a licensor, author or contributor be held liable for indirect,
.\" direct, other damage, loss, or other issues arising in any way out
.\" of dealing in the work, even if advised of the possibility of such
.\" damage or existence of a defect, except proven that it results out
.\" of said person’s immediate fault when using the work as intended.
.\"-
.Dd January 23, 2021
.Dt SLEEP 1
.Os MirBSD
.Sh NAME
.Nm sleep
.Nd suspend execution for an interval of time
.Sh SYNOPSIS
.Nm
.Ar time Op Ar ...
.Sh DESCRIPTION
The
.Nm
utility suspends execution for the amount of time specified by its operands.
.Pp
.Nm
is commonly used to delay proceeding within a series of commands.
Another use is to schedule the execution of other commands, but
.Nm
will not take changes to the system clock that occur during its
execution into account (it sleeps
.Dq for Ar time No seconds ,
not
.Dq until a specific time ) ;
it also uses the simplified notion of a day having 86400 seconds.
.Pp
Each
.Ar time
operand (multiple operands are added together to form a total amount of
time to sleep) must be comprised of a positive number, which may contain
a decimal fraction, and optionally a suffix letter denoting a scaling factor.
The scaling factors supported are as follows:
.Bl -tag -width 1n -offset 2n
.It Ic s
seconds (1); default
.It Ic m
minutes (60)
.It Ic h
hours (3600)
.It Ic d
days (86400)
.El
.Sh EXAMPLES
Monitor the growth of a file without consuming too many resources:
.Pp
.Bd -literal -offset indent -compact
while sleep 5; do
	ls \-l file
done
.Ed
.Pp
Wait an half hour before running the script
.Pa command_file
(see also the
.Xr at 1
utility):
.Pp
.Dl (sleep 1800; sh command_file \*(Gterrors 2\*(Gt&1) &
.Sh DIAGNOSTICS
The
.Nm
utility exits 0 on success or upon receipt of
.Dv SIGALRM ,
and \*(Gt0 if an error occurs.
.Sh SEE ALSO
.Xr at 1 ,
.Xr select 2
.Sh STANDARDS
The
.Nm
utility conforms to
.St -p1003.1-2008 .
.Pp
The handling of fractional arguments is provided as an extension to the
specification.
The handling of multiple arguments and an optional suffix is provided for
compatibility with
.Tn GNU
coreutils'
.Nm
utility.
.Sh AUTHORS
.An mirabilos Aq m@mirbsd.org
.Sh CAVEATS
.Nm
parses fractional arguments up to six digits after the period;
any further digits are silently ignored.
Suffix scaling is performed afterwards and so can deviate up to 86.4 ms.
.Pp
The underlying system call accepts input to the microsecond but may be
using a clock with coarse granularity, which affects the accuracy of
the time actually slept; the
.Nm time
utility is known to use a much less accurate timer, but compiling the
.Nm
utility with
.Ic \-DDEBUG
will show that it always sleeps for at least the requested time.


sleep.c
/*-
 * Copyright © 2021
 *      mirabilos <m@mirbsd.org>
 *
 * Provided that these terms and disclaimer and all copyright notices
 * are retained or reproduced in an accompanying document, permission
 * is granted to deal in this work without restriction, including un‐
 * limited rights to use, publicly perform, distribute, sell, modify,
 * merge, give away, or sublicence.
 *
 * This work is provided “AS IS” and WITHOUT WARRANTY of any kind, to
 * the utmost extent permitted by applicable law, neither express nor
 * implied; without malicious intent or gross negligence. In no event
 * may a licensor, author or contributor be held liable for indirect,
 * direct, other damage, loss, or other issues arising in any way out
 * of dealing in the work, even if advised of the possibility of such
 * damage or existence of a defect, except proven that it results out
 * of said person’s immediate fault when using the work as intended.
 */

#define EXTERN
#include "sleep.h"

__RCSID("$MirOS: src/bin/sleep/sleep.c,v 1.5 2021/01/23 06:12:47 tg Exp $");

#ifdef SMALL
static const char ERR[4] = { 'E', 'R', 'R', '\n' };
#define die(...)		do {		\
	write(2, ERR, sizeof(ERR));		\
	_exit(1);				\
} while (/* CONSTCOND */ 0)
#else
static void die(const char *, ...)
    MKSH_A_NORETURN
    MKSH_A_FORMAT(__printf__, 1, 2);
#endif
#ifdef SIGALRM
static void handler(int)
    MKSH_A_NORETURN;
#endif
static unsigned int classify(const char *);
static int dosleep(time_t, unsigned int);

#ifndef SMALL
static void
die(const char *msg, ...)
{
	va_list ap;

	fputs("E: ", stderr);
	va_start(ap, msg);
	vfprintf(stderr, msg, ap);
	va_end(ap);
	putc('\n', stderr);
	exit(1);
}
#endif

#ifdef SIGALRM
static void
handler(int signo MKSH_A_UNUSED)
{
	_exit(0);
}
#endif

/* 0-9; 10=NUL 11=factor 12=period 13=unknown */
static unsigned int
classify(const char *cp)
{
	switch (ord(*cp)) {
	case ord('\0'):
		return (10);
	case ord('0'):
		return (0);
	case ord('1'):
		return (1);
	case ord('2'):
		return (2);
	case ord('3'):
		return (3);
	case ord('4'):
		return (4);
	case ord('5'):
		return (5);
	case ord('6'):
		return (6);
	case ord('7'):
		return (7);
	case ord('8'):
		return (8);
	case ord('9'):
		return (9);
	case ord('.'):
		return (12);
#ifndef SMALL
	case ord('s'):
	case ord('m'):
	case ord('h'):
	case ord('d'):
		if (1[cp] == '\0')
			return (11);
		/* FALLTHROUGH */
#endif
	default:
		return (13);
	}
}

static int
dosleep(time_t s, unsigned int u)
{
	struct timeval tv;

	tv.tv_sec = s;
	tv.tv_usec = u;
	return (select(0, NULL, NULL, NULL, &tv));
}

int
main(int argc, char *argv[])
{
	int argp;
	time_t sS = 0;
	unsigned int sU = 0;
#ifdef notyet_sleepuntil
	struct timeval t0;
#endif
#ifdef DEBUG
	struct timespec ts1, ts2;

	if (clock_gettime(CLOCK_MONOTONIC, &ts1))
		die("clock_gettime: %s", strerror(errno));
#endif

#ifdef notyet_sleepuntil
#if HAVE_GETTIMEOFDAY
	if (gettimeofday(&t0, NULL))
		die("could not get time of day: %s", strerror(errno));
#else
	if (time(&t0.tv_sec) == (time_t)-1)
		die("could not get time: %s", strerror(errno));
	t0.tv_usec = 0;
#endif
#endif

#ifdef SIGALRM
	signal(SIGALRM, handler);
#endif

	argp = 1 + (argc > 1 &&
	    ord(argv[1][0]) == ord('-') &&
	    ord(argv[1][1]) == ord('-') &&
	    !argv[1][2]);
	if (!(argp < argc))
		die("operand is mandatory");
	while (argp < argc) {
		unsigned int i, j, tU = 0;
		time_t tS = 0;
		const char *cp = argv[argp++];
		char argvalid = 0;
 parse_sec:
		if ((i = classify(cp++)) < 10) {
			if (notoktomula(time_t, tS, 10, i)) {
#ifndef SMALL
 overflow:
#endif
				die("argument too large: %s", argv[--argp]);
			}
			tS = (tS * 10) + i;
			argvalid = 1;
			goto parse_sec;
		} else if (i == 10) {
			goto parse_out;
#ifndef SMALL
		} else if (i == 11) {
			goto parse_factor;
#endif
		} else if (i != 12) {
 invopnd:
			die("invalid char '%c' in operand: %s",
			    *--cp, argv[--argp]);
		}
		/* i == 12 */
		j = 100000;
 parse_usec:
		if ((i = classify(cp++)) < 10) {
			if (j) {
				tU += i * j;
				j /= 10;
				/* we continue to parse digits even past µs */
			}
			argvalid = 1;
			goto parse_usec;
#ifndef SMALL
		} else if (i == 11) {
 parse_factor:
			switch (ord(cp[-1])) {
			default: /* ord('s'): */
				goto parse_out;
			case ord('m'):
				i = 60;
				break;
			case ord('h'):
				i = 3600;
				break;
			case ord('d'):
				i = 86400;
				break;
			}
			if (notoktomula(time_t, tS, i, 0))
				goto overflow;
			tS *= i;
			if (tU) {
				/* split tU .mmmuuu to j=mmm tU=uuu and scale */
				j = (tU / 1000U) * i;
				tU = (tU % 1000U) * i;
				/* add s/ms share of tU to j */
				j += tU / 1000U;
				/* add s share of j to tS */
				i = j / 1000U;
				if (notoktoadd(time_t, tS, i))
					goto overflow;
				tS += i;
				/* reconstruct µs share of tU + ms share of j */
				tU = (tU % 1000U) + ((j % 1000U) * 1000U);
			}
#endif
		} else if (i != 10) {
			goto invopnd;
		}
		/* i == 10 or 11-and-past-parsing-factor */
 parse_out:
		if (!argvalid)
			die("operand with no digits: %s", argv[--argp]);
#ifdef TEST
		printf("add %llu.%06u\n", (unsigned long long)tS, tU);
#endif
		sU += tU;
		if (sU > 999999U) {
			if (notoktoadd(time_t, tS, 1))
				goto sumover;
			++tS;
			sU -= 1000000U;
		}
		if (notoktoadd(time_t, sS, tS)) {
 sumover:
			die("argument sum too large at: %s", argv[--argp]);
		}
		sS += tS;
#ifdef TEST
		printf("sum %llu.%06u\n\n", (unsigned long long)sS, sU);
#endif
	}

	while (sS > (21U * 86400U)) {
		if (dosleep(14U * 86400U, 0)) {
 eintr:
			die("sleep: %s", strerror(errno));
		}
		sS -= 14U * 86400U;
	}
	if (dosleep(sS, sU))
		goto eintr;
#ifdef DEBUG
	if (clock_gettime(CLOCK_MONOTONIC, &ts2))
		die("clock_gettime: %s", strerror(errno));
	timespecsub(&ts2, &ts1, &ts1);
	printf("%llu.%09lu\n", (unsigned long long)ts1.tv_sec, ts1.tv_nsec);
#endif
	return (0);
}
